### ISSUE TITLE
Question : should we add `--opt-level X` flag for benchmarking purpose?

### ISSUE THREAD

**Issue by pramodk** [BlueBrain/nmodl#616]
This is more as a question - for benchmarking via JIT, should we add CLI option to run default optimisation passes like `-O0, -O1, -O2, or -O3`? Seems like [llvm::PassManagerBuilder](https://llvm.org/doxygen/classllvm_1_1PassManagerBuilder.html) has most of the infrastructure to easily add default optimisation passes and it's used in tools like opt, lli (see usage of `populateFunctionPassManager` in [opt.cpp](https://github.com/llvm/llvm-project/blob/6fca189532511da1b48e8c0d9aad8ff2edca382d/llvm/tools/opt/opt.cpp#L416)).

@georgemitenkov : I can add this into NMODL and it might be useful for benchmarking? Before that, I want to check if you saw any reason not to have such option.

**Comment by pramodk** [comment]
Edit : once vector IR is created, existing `--opt` option is running few common passes and it's taking care of most obvious things. Adding `-O2 / -O3` may not awakes bring any additional advantages do but having this interface in place wouldn't hurt ?

**Comment by georgemitenkov** [comment]
>  I can add this into NMODL and it might be useful for benchmarking? 

I agree that for benchmarking having this option would be useful! (If we add `noalias`, `loop` metadata it may yield some more optimisations.) The reason we did not have it is simply because we implemented `--opt` first, and did not prioritise adding the option :)

**Comment by georgemitenkov** [comment]
@pramodk Regarding passes: in LLVM we can specify two:

1. Ones that run on IR (per module/function) (Visitor or pre-JIT level)
2. Ones that are run for code generation (JIT level)

Currently, for the first we used `--opt` and for the second JIT's default (`O2`). I am wondering how do you see `--opt-level` working (with respect to 2 specifications above)?

What I have in mind:
- Keep `--opt` since we may want to experiment with certain passes
- Add `--opt-level x` to `benchmark`, that will run `Ox` for IR (basically adding an extra `transformr` layer in JIT that will operate on IR) and then for codegen passes (JIT level)

**Comment by pramodk** [comment]
> I am wondering how do you see --opt-level working (with respect to 2 specifications above)?

Your proposal looks fine. But, if we are going to benefit by two flags, what if we make naming explicit? e.g. `--opt-level-ir` and `--opt-level-jit`?

This will avoid guessing new person which flag is doing what?

**Comment by georgemitenkov** [comment]
Yes, this can work as well! I guess making my two explicit flags will help to avoid  confusion.
